tests/main_test.ts

```diff
@@ -563,6 +563,21 @@ Deno.test({
   },
 });
 
+Deno.test({
+  name: "/middleware - should pass state through all middlewares",
+  fn: async () => {
+    await withPageName("./tests/fixture/main.ts", async (page, address) => {
+      await page.goto(`${address}/state-middleware/foo`);
+      const text = await page.$eval("pre", (el) => el.textContent);
+      assertEquals(JSON.parse(text), {
+        handler1: "it works",
+        handler2: "it works",
+        handler3: "it works",
+      });
+    });
+  },
+});
+
```

Comment:
I like avoiding `withPageName` since using `handler` makes it way easier to debug. I think we also need to assert that `root_mw` isn't included, since that's overridden when you directly assign state with `ctx.state = { handler1: "it works" }`.
```suggestion
Deno.test({
  name: "/middleware - should pass state through all middlewares",
  fn: async () => {
    const resp = await handler(
      new Request("https://fresh.deno.dev/state-middleware/foo"),
    );
    assert(resp);
    assertEquals(resp.status, Status.OK);
    const body = await resp.text();
    assert(
      !body.includes("root_mw"),
      "shouldn't include root_mw, because we override it at the first layer",
    );
    assertStringIncludes(
      body,
      "&quot;handler1&quot;: &quot;it works&quot;,\n  &quot;handler2&quot;: &quot;it works&quot;,\n  &quot;handler3&quot;: &quot;it works&quot;",
    );
  },
});
```

Reply:
+1

Reply:
Good point about using the handler way to test things.

FYI: `assertEquals` already tests for an exact match of keys and would fail if it included `"root_mw"`.